src/flow_graph.py

```diff
@@ -1096,15 +1253,157 @@ def reset_block_info(self) -> None:
             node.block.block_info = None
 
 
+def phi_reg_sources(node: Node, reg: Access, imdom_srcs: RefSet) -> RefSet:
+    """
+    Return the RefSet of all of the places `reg` is assigned to, if it is a valid phi.
+    Otherwise, return RefSet.invalid().
+    """
+    if node.immediate_dominator is None:
+        return RefSet.invalid()
+
+    seen = {node.immediate_dominator}
+    stack = node.parents[:]
+    sources = RefSet()
+
+    while stack:
+        n = stack.pop()
+        if n == node.immediate_dominator and not imdom_srcs.is_valid():
+            return RefSet.invalid()
+        if n in seen:
+            continue
+        seen.add(n)
+
+        for i, instr in list(enumerate(n.block.instructions))[::-1]:
+            with current_instr(instr):
+                if reg in instr.outputs:
+                    sources.add(InstrRef(n, i))
+                    break
+                if reg in instr.clobbers:
+                    return RefSet.invalid()
+        else:
+            stack.extend(n.parents)
+
+    if not sources.is_valid():
+        return imdom_srcs
+    return sources
```

Comment:
`phi_reg_sources` is based on the old `reg_always_set`, but returns the `RefSet` of the instructions that set the given reg (instead of just a boolean value).

In general, I'm also unsure if it's OK to use the naming convention `reg: Access`? I feel like "register" is a more concrete concept than "access" even if it's technically too specific? All these functions work just as well with `MemoryAccess`s as `Register`s -- I'd be happy to change/redocument anything if you have suggestions for making this clearer.